java-checks/src/main/java/org/sonar/java/checks/SelfAssignementCheck.java

```diff
@@ -19,29 +19,61 @@
  */
 package org.sonar.java.checks;
 
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
 import org.sonar.check.Rule;
+import org.sonar.java.model.DefaultJavaFileScannerContext;
+import org.sonar.java.model.JWarning;
+import org.sonar.java.model.JavaTree;
 import org.sonar.java.model.SyntacticEquivalence;
 import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
 import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.SyntaxToken;
 import org.sonar.plugins.java.api.tree.Tree;
 
-import java.util.Collections;
-import java.util.List;
-
 @Rule(key = "S1656")
 public class SelfAssignementCheck extends IssuableSubscriptionVisitor {
 
+  private static final String ISSUE_MESSAGE = "Remove or correct this useless self-assignment.";
+  private final Set<JWarning> warnings = new HashSet<>();
+  private final Set<JWarning> reportedWarnings = new HashSet<>();
```

Comment:
Instead of having two sets here and subtracting one from the other at the end, it would be simpler to just have the warnings set and remove entries from it as they're reported. This should also improve runtime complexity as we'd no longer loop over all the warnings in `markReportedWarnings` every time a warning is reported.

Reply:
nice finding. I didn't think about this.